File: RDE/data/SDToolbox/Python3/demo/demo_g.py
"""
Shock and Detonation Toolbox Demo Program

This program computes the ratio of specific heats and logarithmic
isentrope slope using several approaches and compares the results
graphically. See Appendix G of the report (below).

################################################################################
Theory, numerical methods and applications are described in the following report:

SDToolbox - Numerical Tools for Shock and Detonation Wave Modeling,
Explosion Dynamics Laboratory, Contributors: S. Browne, J. Ziegler,
N. Bitter, B. Schmidt, J. Lawson and J. E. Shepherd, GALCIT
Technical Report FM2018.001 Revised January 2021.

Please cite this report and the website if you use these routines. 

Please refer to LICENCE.txt or the above report for copyright and disclaimers.

http://shepherd.caltech.edu/EDL/PublicResources/sdt/

################################################################################ 
Updated September 2018
Tested with: 
    Python 3.5 and 3.6, Cantera 2.3 and 2.4
Under these operating systems:
    Windows 8.1, Windows 10, Linux (Debian 9)
"""
import cantera as ct
import matplotlib.pyplot as plt
from sdtoolbox.thermo import gruneisen_eq,gruneisen_fr,soundspeed_eq,soundspeed_fr
import numpy as np

# set composition 
#mech = 'Hong2011.cti'
#mech = 'gri30_highT.cti'
mech = 'Mevel2017.cti'
gas = ct.Solution(mech)
q = 'O2:1 H2:2.001'
T1 = 300
P1 = ct.one_atm
# find species index for plotting
iH = gas.species_index('H')
iO = gas.species_index('O')
iOH = gas.species_index('OH')
iH2O = gas.species_index('H2O')
iH2 = gas.species_index('H2')
iO2 = gas.species_index('O2')

# Use constant volume explosion as starting point for isentropic expansion
gas.TPX = T1,P1,q
gas.equilibrate('UV')
T2 = gas.T
S2 = gas.s
V2 = gas.v
V  = V2
X2 = gas.X

P = []; R = []; T = [] 
xH = []; xO = []; xOH = []; xH2 = []; xO2 = []; xH2O = []
grun_eq = []; grun_fr = []; a_eq = []; a_fr = []; kappa_fr = []; kappa_eq = [];
tvdvdt_eq = []; tvdvdt_fr = []; dpdr = []; gamma_eq = []; gamma_fr = []; 
diff = []; G_mult = []; gamma_fr_thermo = []; gamma_eq_thermo = []; P_f = []; T_f = []; ratio = [];

for i in range(20):
    # equilibrium state computations
    gas.SV = S2,V
    gas.equilibrate('SV')
    P.append(gas.P)
    R.append(gas.density)
    T.append(gas.T)
    # species
    X = gas.X
    xH.append(X[iH])
    xO.append(X[iO])
    xOH.append(X[iOH])
    xH2.append(X[iH2])
    xO2.append(X[iO2])
    xH2O.append(X[iH2O])
    # gamma computed from Gruneisen coefficient
    grun_eq.append(gruneisen_eq(gas))
    grun_fr.append(gruneisen_fr(gas))
    a_eq.append(soundspeed_eq(gas))
    a_fr.append(soundspeed_fr(gas)) 
    # logarithmic slope of isentropes in P-V coordinates
    kappa_fr.append(a_fr[-1]**2*R[-1]/P[-1])
    kappa_eq.append(a_eq[-1]**2*R[-1]/P[-1])
    # logarithmic slope of isentropes in T-V coordinates
    # equilibrium
    gas.SV = S2,1.01*V
    gas.equilibrate('SV')
    T_2 = gas.T
    gas.SV = S2,0.99*V
    gas.equilibrate('SV')
    T_1 = gas.T
    tvdvdt_eq.append(-V*(T_2-T_1)/(0.02*V)/T[-1])
    # frozen
    gas.SVX = S2,V*1.01,X2
    T_2 = gas.T
    gas.SVX = S2,V*0.99,X2
    T_1 = gas.T
    tvdvdt_fr.append(-1.*V*(T_2-T_1)/(0.02*V)/T[-1])
     # compute equilibrium values of isothermal dP/drho
    gas.TD = T[-1],1/(1.01*V)
    gas.equilibrate('TV')
    P2 = gas.P
    gas.TD = T[-1],1/(0.99*V)
    gas.equilibrate('TV')
    P1 = gas.P
    dpdr.append(-V*V*(P2-P1)/(0.02*V))
    # compute ratio of specific heats from thermodynamic identities
    gamma_eq.append(a_eq[-1]*a_eq[-1]/dpdr[-1])
    gamma_fr.append(a_fr[-1]*a_fr[-1]*R[-1]/P[-1])
    # compute the difference between equilbrium and frozen dpdrho_T
    diff.append(dpdr[-1]*R[-1]/P[-1])
    # compute equilibrium values of dvdt_P
    gas.TP = T[-1]*1.01,P[-1]
    gas.equilibrate('TP')
    V_2 = gas.v    
    gas.TP = T[-1]*0.99,P[-1]
    gas.equilibrate('TP')
    V_11 = gas.v    
    dvdt = (V_2-V_11)/(0.02*T[-1])
    G_mult.append(dvdt*T[-1]*R[-1])
    # compute frozen specific heats from thermodynamic properties
    gas.TD = T[-1],1/V
    gas.equilibrate('TV')
    gamma_fr_thermo.append(gas.cp_mass/gas.cv_mass)
    # compute equilibrium specific heats from properties and definition
    gas.TD = T[-1]*1.01,1/V
    gas.equilibrate('TV')
    U2 = gas.u
    gas.TD = T[-1]*0.99,1/V
    gas.equilibrate('TV')
    U1 = gas.u
    CVEQ = (U2-U1)/(.02*T[-1])
    gas.TP = T[-1]*1.01,P[-1]
    gas.equilibrate('TP')
    H2 = gas.h
    gas.TP = T[-1]*0.99,P[-1]
    gas.equilibrate('TP')
    H1 = gas.h
    CPEQ = (H2-H1)/(.02*T[-1])
    gamma_eq_thermo.append(CPEQ/CVEQ)
    ratio.append(gamma_eq[-1]/gamma_eq_thermo[-1])
    # frozen pressure and temperature 
    gas.SVX = S2,V,X2
    P_f.append(gas.P)
    T_f.append(gas.T)
    # next state
    V = 2.*V

# plotting
kappa_fr = np.asarray(kappa_fr); kappa_eq = np.asarray(kappa_eq)
gamma_fr = np.asarray(gamma_fr); gamma_eq = np.asarray(gamma_eq)
gamma_fr_thermo = np.asarray(gamma_fr_thermo); gamma_eq_thermo = np.asarray(gamma_eq_thermo)
R = np.asarray(R); P = np.asarray(P); P_f = np.asarray(P_f); T = np.asarray(T); T_f = np.asarray(T_f);
    
plt.figure(1)
plt.plot(T,grun_fr,':',linewidth=1)
plt.plot(T,kappa_fr-1,'-.',linewidth=1) 
plt.plot(T,gamma_fr-1,'--',linewidth=1)
plt.plot(T,gamma_fr_thermo-1,'-',linewidth=1); 
plt.title(r'Frozen $\gamma$ comparisons',fontsize=12)
plt.xlabel('Temperature (K)',fontsize=12)
plt.ylabel(r'$\gamma - 1$',fontsize=12);
plt.legend([r'$G_{fr}$',r'$\kappa_{fr}-1$',r'$\gamma_{fr}-1$',r'$(c_p/c_v)_{fr} - 1$'])
plt.xlim(min(T),max(T))

plt.figure(2)
plt.plot(T,grun_eq,':',linewidth=1)
plt.plot(T,kappa_eq-1,'-.',linewidth=1)  
plt.plot(T,gamma_eq-1,'--',linewidth=1)
plt.plot(T,gamma_eq_thermo-1,'-',linewidth=1) 
plt.title(r'Equilibrium $\gamma$ comparisons',fontsize=12)
plt.xlabel('Temperature (K)',fontsize=12)
plt.ylabel(r'$\gamma - 1$',fontsize=12)
plt.legend([r'$G_{eq}$',r'$\kappa_{fr}-1$',r'$\gamma_{fr}-1$',r'$(c_p/c_v)_{fr} - 1$'])
plt.xlim(min(T),max(T))

plt.figure(3)
plt.plot(T,gamma_fr-1,':',linewidth=1); 
plt.plot(T,gamma_eq_thermo-1,'-.',linewidth=1)
plt.plot(T,kappa_eq-1,'--',linewidth=1)
plt.title(r'Frozen vs Equilibirum $\gamma$ comparisons',fontsize=12)
plt.xlabel('Temperature (K)',fontsize=12)
plt.ylabel(r'$\gamma - 1$',fontsize=12)
plt.legend([r'$\gamma_{fr}-1$',r'$\gamma_{eq}-1$',r'$\kappa_{eq}-1$'])
plt.xlim(min(T),max(T))

plt.figure(4)
plt.plot(T,diff,'k-',linewidth=1) 
plt.title(r'Equilibrium for $(\rho/P)(dP/d\rho)_T$',fontsize=12)
plt.xlabel('Temperature (K)',fontsize=12)
plt.ylabel(r'$(\rho/P)(dP/d\rho)_T$',fontsize=12)
plt.xlim(min(T),max(T))

plt.figure(5)
plt.plot(T,G_mult,'k-',linewidth=1)
plt.title(r'Equilibrium value of $(T/v)dv/dT_p$',fontsize=12)
plt.xlabel('Temperature (K)',fontsize=12)
plt.ylabel(r'$(T/v)(dv/dT)_P$',fontsize=12)
plt.xlim(min(T),max(T))

plt.figure(6)
plt.plot(T,xH,linewidth=1)
plt.plot(T,xO,linewidth=1)
plt.plot(T,xOH,linewidth=1)
plt.plot(T,xH2,linewidth=1)
plt.plot(T,xO2,linewidth=1) 
plt.plot(T,xH2O,linewidth=1)
plt.yscale('log')
plt.ylim((1e-4,1))
plt.title('Species distribution',fontsize=12)
plt.xlabel('Temperature (K)',fontsize=12)
plt.ylabel('Mole fraction',fontsize=12)
plt.legend(['H','O','OH','H2','O2','H2O'],loc='upper left')
plt.xlim(min(T),max(T))

V_p = 1/R
plt.figure(7)
plt.plot(V_p,T_f,'k--',linewidth=1) 
plt.plot(V_p,T,'k-',linewidth=1)
plt.xscale('log')
plt.yscale('log')
plt.axis((1,1E5,100,4000))
plt.title('Frozen vs Equilibirum isentrope',fontsize=12)
plt.xlabel(r'Volume (m$^3$/kg)',fontsize=12)
plt.ylabel('Temperature (K)',fontsize=12)
plt.legend(['Frozen','Equilibrium'],loc='upper right')

plt.figure(8)
plt.plot(V_p,P_f/ct.one_atm,'k--',linewidth=1) 
plt.plot(V_p,P/ct.one_atm,'k-',linewidth=1)
plt.xscale('log')
plt.yscale('log')
plt.axis((1,1E5,1E-5,20))
plt.title('Frozen vs Equilibirum isentrope',fontsize=12)
plt.xlabel(r'Volume (m$^3$/kg)',fontsize=12)
plt.ylabel('Pressure (atm)',fontsize=12)
plt.legend(['Frozen','Equilibrium'],loc='upper right')


## fitting isentropes
# restrict fit to volumes less than 1000 times initial value
valid = list(np.transpose(np.argwhere(V_p < 1000.*V2))[0])
# first find log values
vlog = np.log(V_p[valid])
telog = np.log(T[valid])
tflog = np.log(T_f[valid])
pelog = np.log(P[valid]/ct.one_atm)
pflog = np.log(P_f[valid]/ct.one_atm)

# fit log with linear fit to get slopes.  For higher accuracy use a 2nd
# order or higher polynomial as there is enough departure from linearity
deg = 1 # polynomial degree 

f = np.poly1d(np.polyfit(vlog,telog,deg))
kfit_te = f[1] # get coefficient of x^1 i.e. slope
x = np.linspace(min(vlog),max(vlog))
plt.figure(9)
plt.plot(x,f(x))
plt.plot(vlog,telog,'o')
plt.title('Equilibrium T-V fit, slope = '+str(kfit_te))
plt.xlabel('log specific volume')
plt.ylabel('log specific temperature')

f = np.poly1d(np.polyfit(vlog,tflog,deg))
kfit_tf = f[1] # get coefficient of x^1 i.e. slope
plt.figure(10)
plt.plot(x,f(x))
plt.plot(vlog,tflog,'o')
plt.title('Frozen T-V fit, slope = '+str(kfit_tf))
plt.xlabel('log specific volume')
plt.ylabel('log specific temperature')

f = np.poly1d(np.polyfit(vlog,pelog,deg))
kfit_pe = f[1] # get coefficient of x^1 i.e. slope
plt.figure(11)
plt.plot(x,f(x))
plt.plot(vlog,pelog,'o')
plt.title('Equilibrium P-V fit, slope = '+str(kfit_pe))
plt.xlabel('log specific volume')
plt.ylabel('log specific temperature')

f = np.poly1d(np.polyfit(vlog,pflog,deg))
kfit_pf = f[1] # get coefficient of x^1 i.e. slope
plt.figure(12)
plt.plot(x,f(x))
plt.plot(vlog,pflog,'o')
plt.title('Frozen P-V fit, slope = '+str(kfit_pf))
plt.xlabel('log specific volume')
plt.ylabel('log specific temperature')

plt.show()

File: RDE/data/SDToolbox/MATLAB/Demo/demo_g.m
% Shock and Detonation Toolbox Demo Program
% 
% This program computes the ratio of specific heats and logarithmic
% isentrope slope using several approaches and compares the results
% graphically. See Appendix G of the report (below).
%  
% ################################################################################
% Theory, numerical methods and applications are described in the following report:
% 
%     Numerical Solution Methods for Shock and Detonation Jump Conditions, S.
%     Browne, J. Ziegler, and J. E. Shepherd, GALCIT Report FM2006.006 - R3,
%     California Institute of Technology Revised September, 2018.
% 
% Please cite this report and the website if you use these routines. 
% 
% Please refer to LICENCE.txt or the above report for copyright and disclaimers.
% 
% http://shepherd.caltech.edu/EDL/PublicResources/sdt/
% 
% ################################################################################ 
% Updated September 2018
% Tested with: 
%     MATLAB 2017b and 2018a, Cantera 2.3 and 2.4
% Under these operating systems:
%     Windows 8.1, Windows 10, Linux (Debian 9)
%%
clear;clc;
disp('demo_g.m Ratio of Specific Heat Evaluation')
% set composition 
%mech = 'Hong2011.cti';
%mech = 'gri30_highT.cti';
mech = 'Mevel2017.cti';
gas = Solution(mech);
q = 'O2:1 H2:2.002'; 
T1 = 300;
P1 = oneatm;
% find species index for plotting
iH = speciesIndex(gas,'H');
iO = speciesIndex(gas,'O');
iOH = speciesIndex(gas,'OH');
iH2O = speciesIndex(gas,'H2O');
iH2 = speciesIndex(gas,'H2');
iO2 = speciesIndex(gas,'O2');
% Use constant volume explosion as starting point for isentropic expansion
set(gas, 'T', T1, 'P', P1, 'X', q);
equilibrate(gas,'UV');
T2 = temperature(gas);
S2 = entropy_mass(gas);
V2 = 1/density(gas);
V  = V2;
X2 = moleFractions(gas);
for i = 1:20
    % equilibrium state computations
    set(gas,'V',V,'S',S2);
    equilibrate(gas,'SV');
    P(i) = pressure(gas);
    R(i) = density(gas);
    T(i) = temperature(gas);
    %species
    X = moleFractions(gas);
    xH(i) = X(iH);
    xO(i) = X(iO);
    xOH(i) = X(iOH);
    xH2(i) = X(iH2);
    xO2(i) = X(iO2);
    xH2O(i) = X(iH2O);
    % gamma computed from Gruneisen coefficient
    grun_eq(i) = gruneisen_eq(gas);
    grun_fr(i) = gruneisen_fr(gas);
    a_eq(i) = soundspeed_eq(gas);
    a_fr(i) = soundspeed_fr(gas); 
    % logarithmic slope of isentropes in P-V coordinates
    kappa_fr(i) =  a_fr(i)*a_fr(i)*R(i)/P(i);
    kappa_eq(i) =  a_eq(i)*a_eq(i)*R(i)/P(i);
    % logarithmic slope of isentropes in T-V coordinates
    % equilibrium
    set(gas,'S',S2,'V',1.01*V)
    equilibrate(gas,'SV');
    T_2 = temperature(gas);
    set(gas,'S',S2,'V',0.99*V)
    equilibrate(gas,'SV');
    T_1 = temperature(gas);
    tvdvdt_eq(i) = -V*(T_2-T_1)/(0.02*V)/T(i);
    % frozen
    set(gas,'S',S2,'V',V*1.01,'X',X2)
    T_2 = temperature(gas);
    set(gas,'S',S2,'V',0.99*V,'X',X2)
    T_1 = temperature(gas);
    tvdvdt_fr(i) = -1.*V*(T_2-T_1)/(0.02*V)/T(i);
     % compute equilibrium values of isothermal dP/drho
    set(gas,'T',T(i),'V',1.01*V)
    equilibrate(gas,'TV');
    P2 = pressure(gas);
    set(gas,'T',T(i),'V',0.99*V)
    equilibrate(gas,'TV');
    P1 = pressure(gas);
    dpdr(i) = -V*V*(P2-P1)/(0.02*V);
    % compute ratio of specific heats from thermodynamic identities
    gamma_eq(i) = a_eq(i)*a_eq(i)/dpdr(i);
    gamma_fr(i) = a_fr(i)*a_fr(i)*R(i)/P(i);
    % compute the difference between equilbrium and frozen dpdrho_T
    diff(i) = dpdr(i)*R(i)/P(i);
    % compute equilibrium values of dvdt_P
    set(gas,'T',T(i)*1.01,'P',P(i));
    equilibrate(gas, 'TP');
    V_2 = 1./density(gas);
    set(gas,'T',T(i)*0.99,'P',P(i));
    equilibrate(gas, 'TP');
    V_11 = 1./density(gas);
    dvdt = (V_2-V_11)/(0.02*T(i));
    G_mult(i) = dvdt*T(i)*R(i);
    % compute frozen specific heats from thermodynamic properties
    set(gas,'T',T(i),'V',V);
    equilibrate(gas, 'TV');
    gamma_fr_thermo(i) = cp_mass(gas)/cv_mass(gas);
    % compute equilibrium specific heats from properties and definition
    set(gas,'T',T(i)*1.01,'V',V);
    equilibrate(gas, 'TV');
    U2 = intEnergy_mass(gas);
    set(gas,'T',T(i)*0.99,'V',V);
    equilibrate(gas, 'TV');
    U1 = intEnergy_mass(gas);
    CVEQ = (U2-U1)/(.02*T(i));
    set(gas,'T',T(i)*1.01,'P',P(i));
    equilibrate(gas, 'TP');
    H2 = enthalpy_mass(gas);
    set(gas,'T',T(i)*0.99,'P',P(i));
    equilibrate(gas, 'TP');
    H1 = enthalpy_mass(gas);
    CPEQ = (H2-H1)/(.02*T(i));
    gamma_eq_thermo(i) = CPEQ/CVEQ;
    ratio(i) = gamma_eq(i)/gamma_eq_thermo(i);
    % frozen pressure and temperature 
    set(gas,'S',S2,'V',V,'X',X2);
    P_f(i) = pressure(gas);
    T_f(i) = temperature(gas);
    % next state
    V = 2.*V;
end
% plotting
	figure(1); clf;
    hold on
	plot(T(:),grun_fr(:),':','LineWidth',1);  
  	plot(T(:),kappa_fr(:)-1,'-.','LineWidth',1);  
    plot(T(:),gamma_fr(:)-1,'--','LineWidth',1); 
    plot(T(:),gamma_fr_thermo(:)-1,'-','LineWidth',1); 
	title(['Frozen \gamma comparisons'],'FontSize', 12,'FontName','Times New Roman');
	xlabel('temperature (K)','FontSize', 12,'FontName','Times New Roman');
	ylabel('\gamma - 1','FontSize', 12,'FontName','Times New Roman');
    legend('G_{fr}','\kappa_{fr}-1','\gamma_{fr}-1','(c_p/c_v)_{fr} - 1');
	set(gca,'FontSize',10,'LineWidth',1);
    hold off
%
	figure(2); clf;
    hold on
	plot(T(:),grun_eq(:),':','LineWidth',1);  
  	plot(T(:),kappa_eq(:)-1,'-.','LineWidth',1);  
    plot(T(:),gamma_eq(:)-1,'--','LineWidth',1); 
    plot(T(:),gamma_eq_thermo(:)-1,'-','LineWidth',1); 
	title(['Equilibrium \gamma comparisons'],'FontSize', 12,'FontName','Times New Roman');
	xlabel('temperature (K)','FontSize', 12,'FontName','Times New Roman');
	ylabel('\gamma - 1','FontSize', 12,'FontName','Times New Roman');
    legend('G_{eq}','\kappa_{eq}-1','\gamma_{eq}-1','(c_p/c_v)_{eq} - 1');
	set(gca,'FontSize',10,'LineWidth',1);
    hold off
    %
	figure(3); clf;
    hold on
    plot(T(:),gamma_fr(:)-1,':','LineWidth',1); 
    plot(T(:),gamma_eq_thermo(:)-1,'-.','LineWidth',1); 
  	plot(T(:),kappa_eq(:)-1,'--','LineWidth',1);  
	title(['Frozen vs Equilibirum \gamma comparisons'],'FontSize', 12,'FontName','Times New Roman');
	xlabel('temperature (K)','FontSize', 12,'FontName','Times New Roman');
	ylabel('\gamma - 1','FontSize', 12,'FontName','Times New Roman');
    legend('\gamma_{fr}-1','\gamma_{eq}-1','\kappa_{eq}-1');
	set(gca,'FontSize',10,'LineWidth',1);
    hold off
    %
	figure(4); clf;
    plot(T,diff,'k-','LineWidth',1); 
	title(['Equilibrium for (\rho/P)(dP/d\rho)_T'],'FontSize', 12,'FontName','Times New Roman');
	xlabel('temperature (K)','FontSize', 12,'FontName','Times New Roman');
	ylabel('(\rho/P)(dP/d\rho)_T','FontSize', 12,'FontName','Times New Roman');
	set(gca,'FontSize',10,'LineWidth',1);
%
	figure(5); clf;
    plot(T,G_mult,'k-','LineWidth',1); 
	title(['Equilibrium value of (T/v)dv/dT_p'],'FontSize', 12,'FontName','Times New Roman');
	xlabel('temperature (K)','FontSize', 12,'FontName','Times New Roman');
	ylabel('(T/v)(dv/dT)_P','FontSize', 12,'FontName','Times New Roman');
	set(gca,'FontSize',10,'LineWidth',1);
%
    figure(6); clf;
    hold on
    plot(T,xH,'LineWidth',1); 
    plot(T,xO,'LineWidth',1); 
    plot(T,xOH,'LineWidth',1'); 
    plot(T,xH2,'LineWidth',1); 
    plot(T,xO2,'LineWidth',1); 
    plot(T,xH2O,'LineWidth',1); 
 	title(['Species distribution'],'FontSize', 12,'FontName','Times New Roman');
	xlabel('temperature (K)','FontSize', 12,'FontName','Times New Roman');
	ylabel('mole fraction','FontSize', 12,'FontName','Times New Roman');
    legend('H','O','OH','H2','O2','H2O','Location','northwest');
	set(gca,'FontSize',10,'LineWidth',1,'Yscale','log','Ylim',[1E-4 1],'YMinorTick','on');
    hold off
    %
    V_p = 1./R;
	figure(7); clf;
    hold on
    plot(V_p,T_f,'k--','LineWidth',1); 
    plot(V_p,T,'k-','LineWidth',1); 
  	title(['Frozen vs Equilibirum isentrope'],'FontSize', 12,'FontName','Times New Roman');
	xlabel('volume (m^3/kg)','FontSize', 12,'FontName','Times New Roman');
	ylabel('temperature (K)','FontSize', 12,'FontName','Times New Roman');
    legend('Frozen','Equilibrium','Location','northeast');
	set(gca,'FontSize',10,'LineWidth',1,'Yscale','log','Ylim',[100 4000],'YMinorTick','on','Xscale','log','Xlim',[1 1E5],'XMinorTick','on');
    hold off
    %
    figure(8); clf;
    hold on
    plot(V_p,P_f/oneatm,'k--','LineWidth',1); 
    plot(V_p,P/oneatm,'k-','LineWidth',1); 
  	title(['Frozen vs Equilibirum isentrope'],'FontSize', 12,'FontName','Times New Roman');
	xlabel('volume (m^3/kg)','FontSize', 12,'FontName','Times New Roman');
	ylabel('pressure (atm)','FontSize', 12,'FontName','Times New Roman');
    legend('Frozen','Equilibrium','Location','northeast');
	set(gca,'FontSize',10,'LineWidth',1,'Yscale','log','Ylim',[1E-5 20],'YMinorTick','on','Xscale','log','Xlim',[1 1E5],'XMinorTick','on');
    hold off
    %
    %% fitting isentropes
    % restrict fit to volumes less than 1000 times initial value
    valid = find(V_p < 1000.*V2);
    last = length(valid);
    % first find log values
    vlog = log(V_p(1:last));
    telog = log(T(1:last));
    tflog = log(T_f(1:last));
    pelog = log((P(1:last)/oneatm));
    pflog = log((P_f(1:last)/oneatm));
    % fit log with linear fit to get slopes.  For higher accuracy use a 2nd
    % order or higher polynomial as there is enough departure from
    % linearity 
    f = fit(vlog',telog','poly1');
    kfit_te = f.p1;
    figure(9); clf
    plot(f,vlog,telog,'o');
    title(['Equilibrium T-V fit, slope = ', num2str(kfit_te)]);
    xlabel('log specific volume');
    ylabel('log specific temperature');
    %
    f = fit(vlog',tflog','poly1');
    kfit_tf = f.p1;
    figure(10); clf
    plot(f,vlog,tflog,'o')
    title(['Frozen T-V fit, slope = ', num2str(kfit_tf)]);
    xlabel('log specific volume');
    ylabel('log specific temperature');
    %
    f = fit(vlog',pelog','poly1');
    kfit_pe = f.p1;
    figure(11); clf
    plot(f,vlog,pelog,'o')
     title(['Equilibrium P-V fit, slope = ', num2str(kfit_pe)]);
    xlabel('log specific volume');
    ylabel('log specific temperature');
   %
    f = fit(vlog',pflog','poly1');
    kfit_pf = f.p1;
    figure(12); clf
    plot(f,vlog,pflog,'o')
     title(['Frozen P-V fit, slope = ', num2str(kfit_pf)]);
    xlabel('log specific volume');
    ylabel('log specific temperature');

